rust/otap-dataflow/crates/otap/src/otap_batch_processor.rs

```diff
@@ -0,0 +1,922 @@
+    /// Go behavior: 0 (or missing) => use send_batch_size; we use 0 as default and normalize later
+    #[serde(default = "default_send_batch_max_size")]
+    pub send_batch_max_size: usize,
+    /// Flush non-empty batches on this interval (milliseconds).
+    #[serde(default = "default_timeout_ms")]
+    pub timeout: u64,
+    /// Optional metadata partitioning keys (resource/scope/attribute names)
+    #[serde(default)]
+    pub metadata_keys: Vec<String>,
+    /// Optional limit on the number of distinct metadata-based groups this processor will track.
+    ///
+    /// Note: This is currently a no-op because grouping by metadata_keys has not yet been
+    /// implemented. When grouping lands, this will cap the number of concurrent groups and the
+    /// overflow strategy will be documented.
+    #[serde(default)]
+    pub metadata_cardinality_limit: Option<usize>,
+}
+
+fn default_send_batch_size() -> usize {
+    DEFAULT_SEND_BATCH_SIZE
+}
+
+fn default_send_batch_max_size() -> usize {
+    FOLLOW_SEND_BATCH_SIZE_SENTINEL // normalized in from_config
+}
+
+fn default_timeout_ms() -> u64 {
+    DEFAULT_TIMEOUT_MS
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            send_batch_size: default_send_batch_size(),
+            send_batch_max_size: default_send_batch_max_size(),
+            timeout: default_timeout_ms(),
+            metadata_keys: Vec::new(),
+            metadata_cardinality_limit: None,
+        }
+    }
+}
+
+/// Local (!Send) OTAP batch processor
+pub struct OtapBatchProcessor {
+    config: Config,
+    // Per-signal buffers of convertible OTAP records
+    current_logs: Vec<OtapArrowRecords>,
+    current_metrics: Vec<OtapArrowRecords>,
+    current_traces: Vec<OtapArrowRecords>,
+    // For inputs that cannot be converted to OTAP (currently dropped with a log)
+    passthrough: Vec<OtapPdata>,
+    // Running row counts per signal for size triggers
+    rows_logs: usize,
+    rows_metrics: usize,
+    rows_traces: usize,
+}
+
+impl OtapBatchProcessor {
+    /// Construct a processor wrapper from a JSON configuration object and processor runtime config.
+    /// The JSON should mirror the Go collector batchprocessor shape. Missing fields fall back to
+    /// crate defaults. Invalid numeric values (e.g., zero) are normalized to minimal valid values.
+    pub fn from_config(
+        node: NodeId,
+        cfg: &Value,
+        proc_cfg: &ProcessorConfig,
+    ) -> Result<ProcessorWrapper<OtapPdata>, ConfigError> {
+        // Allow Go-style timeout strings; parse to milliseconds before deserialization.
+        let mut cfg_norm = cfg.clone();
+        if let Some(s) = cfg.get("timeout").and_then(|v| v.as_str()) {
+            if let Some(ms) = parse_duration_ms(s) {
+                if let Some(obj) = cfg_norm.as_object_mut() {
+                    let _ = obj.insert("timeout".to_string(), Value::from(ms));
+                }
+            }
+        }
```

Comment:
Do we need to support both timeout as millisecond and parsing in human readable format? I think it might just be better to support parsing from human readable to simplify things.

You could change the definition of `timeout` in the config object to:
```rs
    /// Flush non-empty batches on this interval.
    #[serde(with = "humantime_serde")]
    pub timeout: Option<Duration>,
```
And then the value would automaticaly get parsed into a `std::time::Duration` when you call `serde_json` below.